test/t/unit/test_unit_dequote.py

```diff
@@ -0,0 +1,161 @@
+            bash, "__tester \\'abc\\'\\\\\\'\\'def\\'", want_output=True
+        )
+        assert output.strip() == "<abc'def>"
+
+    def test_8_param_1(self, bash):
+        output = assert_bash_exec(bash, "__tester '$v'", want_output=True)
+        assert output.strip() == "<var>"
+
+    def test_8_param_2(self, bash):
+        output = assert_bash_exec(bash, "__tester '${v}'", want_output=True)
+        assert output.strip() == "<var>"
+
+    def test_8_param_3(self, bash):
+        output = assert_bash_exec(bash, "__tester '${#v}'", want_output=True)
+        assert output.strip() == "<3>"
+
+    def test_8_param_4(self, bash):
+        output = assert_bash_exec(bash, "__tester '${v[0]}'", want_output=True)
+        assert output.strip() == "<var>"
+
+    def test_9_qparam_1(self, bash):
+        output = assert_bash_exec(bash, "__tester '\"$v\"'", want_output=True)
+        assert output.strip() == "<var>"
+
+    def test_9_qparam_2(self, bash):
+        output = assert_bash_exec(
+            bash, "__tester '\"${v[@]}\"'", want_output=True
+        )
+        assert output.strip() == "<var>"
+
+    def test_10_pparam_1(self, bash):
+        output = assert_bash_exec(bash, "__tester '$?'", want_output=True)
+        assert output.strip() == "<0>"
+
+    def test_10_pparam_2(self, bash):
+        output = assert_bash_exec(bash, "__tester '${#1}'", want_output=True)
+        assert output.strip() == "<5>"  # The string `${#1}` is five characters
+
+    def test_unsafe_1(self, bash):
+        output = assert_bash_exec(
+            bash, "! __tester '$(echo hello >&2)'", want_output=True
+        )
+        assert output.strip() == ""
+
+    def test_unsafe_2(self, bash):
+        output = assert_bash_exec(
+            bash, "! __tester '|echo hello >&2'", want_output=True
+        )
+        assert output.strip() == ""
+
+    def test_unsafe_3(self, bash):
+        output = assert_bash_exec(
+            bash, "! __tester '>| important_file.txt'", want_output=True
+        )
+        assert output.strip() == ""
+
+    def test_unsafe_4(self, bash):
+        output = assert_bash_exec(
+            bash, "! __tester '`echo hello >&2`'", want_output=True
+        )
+        assert output.strip() == ""
+
+    def test_glob_default(self, bash):
+        with bash_env_saved(bash) as bash_env:
+            bash_env.shopt("failglob", False)
+            bash_env.shopt("nullglob", False)
+            output = assert_bash_exec(
+                bash, "__tester 'non-existent-*.txt'", want_output=True
+            )
+            assert output.strip() == "<non-existent-*.txt>"
+
+    # def test_glob_noglob(self, bash):
+    #     with bash_env_saved(bash) as bash_env:
+    #         bash_env.set("noglob", True)
+    #         output = assert_bash_exec(
+    #             bash,
+    #             "__tester 'non-existent-*.txt'",
+    #             want_output=True,
+    #         )
+    #         assert output.strip() == "<non-existent-*.txt>"
```

Comment:
This test is currently commented out because it requires f0a4028d87f3506f2091dbafef91771794632c94 contained in #708.
- (a) #708 may be processed before this one, or
- (b) this code section may be uncommented after merging #708, or
- (c) f0a4028 and related commits may be cherry-picked into this PR.

Reply:
Cool, let's finalize #708 first, I just left some last comments there.

Reply:
I have uncommented it (8c3ae045) and rebased/squashed it.